keps/sig-storage/3141-prevent-volume-mode-conversion/README.md

```diff
@@ -0,0 +1,761 @@
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+<!--
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+-->
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+With the introduction of the VolumeSnapshot feature in Kubernetes 1.20, a
+PersistentVolumeClaim (PVC) can be restored from a previously taken VolumeSnapshot.
+This is done by pointing the “dataSource” parameter of the PVC spec to an existing
+VolumeSnapshot resource. There is no logic that validates whether the original
+volume mode of the PVC, whose snapshot was taken, matches the volume mode of 
+the newly created PVC that is being restored from the existing VolumeSnapshot.
+
+## Motivation
+
+Malicious users may exploit a vulnerability in the kernel by exploiting
+this gap. However we cannot simply block this operation as some backup
+vendors try to restore a volume with the exact same mode as the original
+volume but may need to do the conversion for efficiency.
+
+### Goals
+
+Define a storage related security standard to mitigate the issue without
+hampering valid use cases.
+
+### Non-Goals
+
+Design that is generic and can be extended to other storage related security
+aspects. This proposal only addresses the snapshot restore security issue.
+
+## Proposal
+
+Here is an example of how a malicious user can exploit this gap to crash the
+kernel.
+1. User creates a PVC with “volumeMode: Block” and runs a pod with it.
+2. User writes malformed ext4 data to it (simple dd)
+3. User takes snapshot of this volume.
+4. User restores snapshot as a PVC with “volumeMode: Filesystem”
+5. User uses this PVC in a pod. 
+   1. kubelet tries to mount it during pod creation. If there is a CVE in the 
+   kernel, the user can crash it.
+
+We cannot simply block this operation as it is a valid use case in for some 
+backup vendors.
+
+The proposal aims to mitigate this issue by modifying the `VolumeSnapshotContent`
+API spec as well as the control flows of `snapshot-controller` and `external-provisioner`.
+`VolumeSnapshotContent` API will include a field that denotes the volume mode of
+the volume that the snapshot was created from. 
+This proposal also introduces a new annotation on the `VolumeSnapshotContent` resource
+that a trusted user (like a backup software) needs to apply on a VolumeSnapshot.
+
+### User Stories (Optional)
+
+<!--
+Detail the things that people will be able to do if this KEP is implemented.
+Include as much detail as possible so that people can understand the "how" of
+the system. The goal here is to make this feel real for users without getting
+bogged down.
+-->
+
+#### Story 1
+
```

Comment:
Can you add details of how a backup software may need to convert volume mode during backup/restore in the user story section?

Reply:
Done, PTAL!